### PR TITLE
[BugFix] Fix serveral complex type bugs in parquet reader

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## Why I'm doing:
Fix below three bugs:

1. `select * from tbl where col1.sub = 10`, we shouldn't prune column col1 for this SQL.
2. If the parquet map's key is a complex type, the parquet reader will throw `ERROR 1064 (HY000): Remaining 5 unparsed field: file = oss:xxxx` error msg. I think we have to throw more clear error msg.
3. If a table is `col STRUCT<sub1: STRUCT<str: STRING>>`, and it's row value is NULL. `SELECT col.sub1.str FROM tbl` should return `NULL`, not `''`.

## What I'm doing:

Fix it

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by zombee0 — APPROVED** [review]

**Review by dirtysalt — DISMISSED** [review]

File: be/src/exprs/array_element_expr.cpp [hunk 1]
```diff
@@ -35,7 +35,8 @@ class ArrayElementExpr final : public Expr {
 
     StatusOr<ColumnPtr> evaluate_checked(ExprContext* context, Chunk* chunk) override {
         DCHECK_EQ(2, _children.size());
-        DCHECK_EQ(_type, _children[0]->type().children[0]);
+        // After DLA's complex type prune, ArrayElement expr's type is different from children's type
```
Comment (trueeyu) [inline be/src/exprs/array_element_expr.cpp:38]: If the types are inconsistent, should cast first？

File: be/src/exprs/subfield_expr.cpp [hunk 2]
```diff
@@ -39,9 +39,23 @@ class SubfieldExpr final : public Expr {
...
+        std::vector<uint8_t> null_flags;
+        size_t num_rows = col->size();
+        null_flags.resize(num_rows, false);
+
+        for (size_t i = 0; i < _used_subfield_names.size(); i++) {
+            const std::string& fieldname = _used_subfield_names[i];
+
+            // merge null flags for each level
+            if (col->is_nullable()) {
+                auto* nullable = down_cast<NullableColumn*>(col.get());
```
Comment (trueeyu) [inline be/src/exprs/subfield_expr.cpp:52]: Why not use FunctionHelper::union_nullable_column or FunctionHelper::union_null_column?

File: be/src/exprs/subfield_expr.cpp [hunk 3]
```diff
@@ -39,9 +39,23 @@ class SubfieldExpr final : public Expr {
...
+        // handle nullable column
+        std::vector<uint8_t> null_flags;
+        size_t num_rows = col->size();
+        null_flags.resize(num_rows, false);
+
+        for (size_t i = 0; i < _used_subfield_names.size(); i++) {
+            const std::string& fieldname = _used_subfield_names[i];
+
+            // merge null flags for each level
+            if (col->is_nullable()) {
```
Comment (trueeyu) [inline be/src/exprs/subfield_expr.cpp:51]: Could it be a `ConstNullColumn` here?

File: be/src/exprs/subfield_expr.cpp (hunk 2 above)
Comment (Smith-Cruise) [reply be/src/exprs/subfield_expr.cpp:52]: fixed

File: be/src/exprs/subfield_expr.cpp (hunk 3 above)
Comment (Smith-Cruise) [reply be/src/exprs/subfield_expr.cpp:51]: it can't

**Review by trueeyu — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**